### ISSUE TITLE
CMake config support

### ISSUE THREAD

**Issue by Xeverous** [cnjinhao/nana#438]
There is no `nana-config.cmake` or `nanaConfig.cmake` generated or installed. The library can be built with CMake (I can `cmake .. params... && make`), but it can not be *used* with CMake (in the form of `find_package(nana)` in any other project) since there is no config. The library has an install script, but it copies only compiled objects and headers.

**Comment by qPCR4vir** [comment]
This can be made, but I unerstand that the idea of `find_package(nana)` if for libraries that dont have direct sopourt  for cmake, and some 3-party team develops that `find_package(nana)`. When the library have "native" cmake you simple "include" it in your project and link with nana, nothing else - much easer.

**Comment by Xeverous** [comment]
> I understand that the idea of `find_package(nana)` is for libraries that don't have direct sopourt for cmake, and some 3-party team develops that `find_package(nana)`.

No. Every library that wants to build with CMake should have it's own `name-config.cmake` so that users are not forced to `add_subdirectory`. Adding subdirectories requires more file operations, does not allow for easy system-wide install and may easily break if a project has multiple dependencies, and say, each of these dependencies want the same library.

`add_subdirectory` is not bad in itself, but it creates a mess in the repository and forces certain project structure which might be very undesirable, especially if the library is required by multiple projects.

`add_subdirectory` is OK if you are making your own project and your own library is used only by your executable. But if you want the library to be used by others you should not require them to add the subdirectory.

> When the library have "native" cmake you simple "include" it in your project and link with it, nothing else - much easer.

No. Imagine I had to `include(Boost)` in every of my projects. I want 1 system-wide installation that is used by all projects.

___

If you want, I can work on this `nana-config.cmake`.

**Comment by qPCR4vir** [comment]
Boost is the bad example: Boost is not built using CMake, you can not add_subdirectory( .. boost ..)

**Comment by qPCR4vir** [comment]
No. If you are forzed to use precompiled libraries you need find_package. 
But if you have the adventage of compiling the libraies from source with your projects **that is by far the best situation** and in this case `add_subdirectory `is what you need, far easer to use. You will alway need to point to where are the libraries anyway, so that is not a major problem. 
The real problem is to get the libaries compiled full compatible with your project. 
system-wide install of precompiled libraries?? I have 4 vesions of gcc in my windows 7 PC. 3 come with precompiled Boost (adding about 500MB), each "system-wide installed". The funy thing is that the 4th gcc installed can not use any of the other 3 (not ABI compatibles, different MinGW distributions). Is not better to have about 5 MB of source code and compile the library with your project, garanting is compatible? Every project include that library using the same binary cmake directory, so no duplication or recompilation needed if they are compatible.
The external libraries are outside of your directory, not any mess here, no forved directory structure. And if just want, you can use `find_file`, `find_library `or similar.

**Comment by qPCR4vir** [comment]
https://youtu.be/rLopVhns4Zs?t=1787

Anyway, people want find_package and want to distribute binaries (I prefer source, when I can compile). Nana distribute sources, but some have create nana pgk. It can be good to have FindNana too.

**Comment by qPCR4vir** [comment]
Sorry, I just wanted to be clear  for future readers:

> The library can be built with CMake (I can cmake .. params... && make), but it can not be used with CMake (in the form of find_package(nana) in any other project) since there is no config. 

This is wrong. The library can be used with cmake, using `add_subdirectory ` with is the recommended way. And can be used `with find_package(nana)` as soon as some 3-party creates and support it, which is the normal way.

**Comment by qPCR4vir** [comment]
See this: https://cliutils.gitlab.io/modern-cmake/modern-cmake.pdf , pag 49.

> Exporting and Installing
There are three good ways and one bad way to allow others use your library:

> ### Find module (the bad way)
> **If you are the library author, don't make a Find<mypackage>.cmake script!** These were designed for libraries _whose authors did not support CMake_. Use a Config<mypackage>.cmake instead as listed below.
> ## Add Subproject
> A package can include your project in a subdirectory, and then use `add_subdirectory`on the subdirectory. This useful for header-only and quick-to-compile libraries. Note that the install commands may interfere with the parent project, so you can add `EXCLUDE_FROM_ALL `to the `add_subdirectory `command; the targets you explicitly use will still be built.
> In order to support this as a library author, make sure you use `CMAKE_CURRENT_SOURCE_DIR `instead of `PROJECT_SOURCE_DIR `(and likewise for other variables, like binary dirs). You can check `CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME` to only add options or defaults that
> make sense if this is a project.
> Also, since namespaces are a good idea, and the usage of your library should be consistent with the other methods below, you should add
> `add_library(MyLib::MyLib ALIAS MyLib)`
> to standardise the usage across all methods. This ALIAS target will not be exported below.

...

**Comment by Xeverous** [comment]
Ok, I watched the videos however I have not seen any significant recommendation for `add_subdirectory`. But I learned that `add_subdirectory` does not have to take an actual subdirectory.

I tried this approach, this is a part of my CMake lists:

```
if(NOT DEFINED NANA_REPO_DIR OR NOT DEFINED NANA_BUILD_DIR)
	message(SEND_ERROR "Define NANA_REPO_DIR to point to nana's GUI library repository and NANA_BUILD_DIR to nana's build directory")
else()
	add_subdirectory(${NANA_REPO_DIR} ${NANA_BUILD_DIR})
endif()
```

CMake documentation says that if `add_subdirectory` gets an absolute path, I also must to specify the path where build is placed. I specified both paths as an absolute and got this error:

```
CMake Error at src/gui/CMakeLists.txt:5 (add_subdirectory):
  The source directory

    path/to/nana/build/directory

  does not contain a CMakeLists.txt file.
```

I don't understand why CMake is trying to find a CMake list in build directory.

**Comment by qPCR4vir** [comment]
What is the value of NANA_REPO_DIR?

**Comment by Xeverous** [comment]
wait lol nvm I had a typo in the path, it works now but apparently you do not export an alias:

```
target [...] links to target "nana::nana" but the target was
  not found.  Perhaps a find_package() call is missing for an IMPORTED
  target, or an ALIAS target is missing?
```

**Comment by Xeverous** [comment]
`nana` works but CMake guides recommend to export a target in a namespace

**Comment by qPCR4vir** [comment]
No, by now we link just with nana:
`target_link_libraries(your_executable PUBLIC nana)`
me too... have a lot to learn, but that have nothing to do with find_package()

**Comment by qPCR4vir** [comment]
> `nana` works but CMake guides recommend to export a target in a namespace

yes, but this is once again - just to offer compatibility with people using find_package(), which by now nana dont provide anyway... nana is too modern for that ;-)

**Comment by Xeverous** [comment]
It is not "just for the package"

See https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#using-a-library-defined-in-the-same-cmake-tree-should-look-the-same-as-using-an-external-library

The name speaks for itself. It doesn't matter how the library is found (`find_package`, `add_subdirectory` or directly in the current project) - it's usage should be just like anythting external. If you provide an alias you will not break anything, allow more library components and idiomaric usage.
